lib/src/transaction/signable.ts

```diff
@@ -47,29 +69,152 @@ export class SignableTransaction {
      */
     public constructor(params: SignableTransactionParams) {
         ow(params, 'params', owSignableTransactionParams);
+        this.network = params.network;
+
+        const cosmosTxDecoded: CosmosTx = JSON.parse(params.rawTxJSON);
+
+        let cosmosObj = cosmosTxDecoded;
 
-        if (params.txBody.value.messages.length === 0) {
-            throw new TypeError('Expected message in `txBody` of `params`, got none');
+        if (cosmosObj.tx) {
+            cosmosObj = cosmosObj.tx;
         }
-        if (params.authInfo.signerInfos.length === 0) {
-            throw new TypeError('Expected signer in `signerInfos` of `authInfo` of `params`, got none');
+
+        if (!cosmosObj.body) {
+            throw new Error('Missing body in Cosmos JSON');
         }
-        if (params.signerAccounts.length === 0) {
+        const { body } = cosmosObj;
+        const { memo } = body;
+        const timeoutHeight = body.timeout_height;
+
+        // TODO: If extension_options and non_critical_extension_options length > 0, then throw
+        if (
+            (body.non_critical_extension_options && body.non_critical_extension_options.length > 0) ||
+            (body.extension_options && body.extension_options.length > 0)
+        ) {
+            throw new Error("SignableTransaction doesn't support 'nonCriticalExtensionOptions' or 'extensionOptions'");
+        }
+
+        if (!body.messages || body.messages.length < 1) {
+            throw new Error('Decoded TxBody does not have valid messages');
+        }
+        const croSdk = CroSDK({ network: this.getNetwork() });
+
+        const txBody: TxBody = {
+            typeUrl: '/cosmos.tx.v1beta1.TxBody',
+            value: {
+                messages: [],
+                memo,
+                timeoutHeight,
+            },
+        };
+
+        body.messages.forEach((message) => {
+            const msgClassInstance = typeUrlToMsgClassMapping(croSdk, message['@type']);
+            const nativeMsg: CosmosMsg = msgClassInstance.fromCosmosMsgJSON(JSON.stringify(message), this.getNetwork());
+            txBody.value.messages.push(nativeMsg);
+        });
+
+        // TODO: structure integrity check
+        const cosmosAuthInfo = cosmosObj.auth_info;
+        const cosmosSignerInfos = cosmosAuthInfo.signer_infos;
+        const signerInfos: SignerInfo[] = [];
+
+        cosmosSignerInfos.forEach((signerInfo) => {
+            // TODO: Support MultiSig in near future
+            const publicKeyObj = signerInfo.public_key as any;
+            if (!publicKeyObj.key) {
```

Comment:
Here it hurts a bit on the readability a bit, can we do?
```suggestion
            if (!signerInfo.mode_info.single) {
```

Reply:
because not having `publicKeyObj.key` does not imply it is a single sign if you do not have knowledge of how multisig Cosmos JSON looks like

Reply:
Makes sense.
 I have made changes.
 resolved.